File: abstract-factory/typescript/custom-operation.ts
import {Connection} from "./connection";
import {Command} from "./command";
import {IDatabaseFactory} from "./database-factory.interface";

export class CustomOperation<T extends IDatabaseFactory> {
  private connection: Connection;
  private command: Command;

  constructor(private readonly databaseFactory: T) {
    this.connection = databaseFactory.createConnection();
    this.command = databaseFactory.createCommand();
  }

  public removeById(id: number): void {
    this.connection.openConnection();
    this.command.executeCommand('DELETE ...');
    this.connection.closeConnection();
  }
}

File: abstract-factory/dart/custom_operation.dart
import 'command.dart';
import 'connection.dart';
import 'i_database_factory.dart';

final class CustomOperation {
  late final IDatabaseFactory _databaseFactory;
  late final Connection _connection;
  late final Command _command;

  CustomOperation(IDatabaseFactory databaseFactory) {
    _databaseFactory = databaseFactory;
    _command = _databaseFactory.createCommand();
    _connection = _databaseFactory.createConnection();
  }

  void removeById(int id) {
    _connection.openConnection();
    _command.executeCommand('DELETE ...');
    _connection.closeConnection();
  }
}
